src/helpers/gateway/mod.rs

```diff
@@ -40,8 +40,9 @@ pub type ReceivingEnd<M> = ReceivingEndBase<TransportImpl, M>;
 pub struct Gateway<T: Transport = TransportImpl> {
     config: GatewayConfig,
     transport: RoleResolvingTransport<T>,
-    senders: GatewaySenders,
-    receivers: GatewayReceivers<T>,
+    senders: Arc<Mutex<GatewaySenders>>,
```

Comment:
both `GatewaySenders` and `GatewayReceivers` are Send + Sync, so they don't need to be protected by a mutex. You could just slap `#derive(Clone)` on them

Reply:
Got it. That's much more elegant.